syntax = "proto3";

service Medusa {
    rpc Backup(BackupRequest) returns (BackupResponse);

    rpc BackupStatus(BackupStatusRequest) returns (BackupStatusResponse);

    rpc DeleteBackup(DeleteBackupRequest) returns (DeleteBackupResponse);

    rpc GetBackups(GetBackupsRequest) returns (GetBackupsResponse);
}

message BackupRequest {
    string name = 1;
    enum Mode {
        DIFFERENTIAL = 0;
        FULL = 1;
    }
    Mode mode = 2;
}

message BackupResponse {

}

message BackupStatusRequest {
    string backupName = 1;
}

message BackupStatusResponse {
    repeated string finishedNodes = 1;
    repeated string unfinishedNodes = 2;
    repeated string missingNodes = 3;
    string startTime = 4;
    string finishTime = 5;
}

message DeleteBackupRequest {
    string name = 1;
}

message DeleteBackupResponse {
}

message GetBackupsRequest {

}

message GetBackupsResponse {
    repeated BackupSummary backups = 1;
}

message BackupSummary {
    string backupName = 1;
    int64 startTime = 2;
    int64 finishTime = 3;
    int32 totalNodes = 4;
    int32 finishedNodes = 5;
    repeated BackupNode nodes = 6;
}

message BackupNode {
    string host = 1;
    repeated int64 tokens = 2;
    string datacenter = 3;
    string rack = 4;
}